File: src/main/java/ludwig/interpreter/Closure.java
package ludwig.interpreter;

import ludwig.model.*;
import ludwig.utils.PrettyPrinter;
import org.pcollections.HashPMap;

public class Closure implements Callable {
    private final HashPMap<NamedNode, Object> locals;
    private final LambdaNode lambda;
    private int argCount;

    public Closure(HashPMap<NamedNode, Object> locals, LambdaNode lambda) {
        this.locals = locals;
        this.lambda = lambda;

        for (int i = 0; i < lambda.children().size(); i++) {
            Node node = lambda.children().get(i);
            if (!(node instanceof VariableNode)) {
                argCount = i;
                break;
            }
        }
    }

    @Override
    public Object tail(Object[] args) {
        HashPMap<NamedNode, Object> env = locals;
        Evaluator visitor = null;
        Object result = null;

        for (int i = 0; i < lambda.children().size(); i++) {
            Node node = lambda.children().get(i);
            if (node instanceof VariableNode) {
                env = env.plus((NamedNode) node, args[i]);
            } else {
                if (visitor == null) {
                    visitor = new Evaluator(env);
                }
                result = node.accept(visitor);
                if (result instanceof Signal) {
                    break;
                }
            }
        }

        return result;
    }

    @Override
    public int argCount() {
        return argCount;
    }

    @Override
    public String toString() {
        return PrettyPrinter.Companion.print(lambda);
    }
}

File: src/main/java/ludwig/interpreter/Closure.kt
package ludwig.interpreter

import ludwig.model.*
import ludwig.utils.PrettyPrinter
import org.pcollections.HashPMap

class Closure(private val locals: HashPMap<NamedNode<*>, Any>, private val lambda: LambdaNode) : Callable {
    private var argCount: Int = 0

    init {

        for (i in 0..lambda.children().size - 1) {
            val node = lambda.children()[i]
            if (node !is VariableNode) {
                argCount = i
                break
            }
        }
    }

    override fun tail(args: Array<Any?>): Any? {
        var env: HashPMap<NamedNode<*>, Any> = locals
        var visitor: Evaluator? = null
        var result: Any? = null

        for (i in 0..lambda.children().size - 1) {
            val node = lambda.children()[i]
            if (node is VariableNode) {
                env = env.plus(node as NamedNode<*>, args[i])
            } else {
                if (visitor == null) {
                    visitor = Evaluator(env)
                }
                result = node.accept(visitor)
                if (result is Signal) {
                    break
                }
            }
        }

        return result
    }

    override fun argCount(): Int {
        return argCount
    }

    override fun toString(): String {
        return PrettyPrinter.print(lambda)
    }
}
